rstream/client.py

```diff
@@ -461,14 +476,50 @@ def __init__(
         self._clients: dict[Addr, Client] = {}
 
     async def get(self, addr: Optional[Addr] = None) -> Client:
-        if addr is None:
-            addr = self.addr
+        """Get a client according to `addr` parameter
+
+        If class param `address_resolver` is set and enabled, load balancer
+        logic will be excuted. Here we do not attempt to connect to the desired `addr`
+        directly, but rather connect to the address defined in the `address_resolver` in a loop
+        until the desired client is returned.
+        """
+        desired_addr = addr or self.addr
+
+        if desired_addr not in self._clients:
+            # don't resolve the client when addr is None. Use the default address
+            # as this is likely a request to get metadata and the node isn't important
+            if addr and self.address_resolver and self.address_resolver.enabled:
+                self._clients[desired_addr] = await self.resolve_client(desired_addr)
+            else:
+                self._clients[desired_addr] = await self.new(desired_addr)
+
+        assert self._clients[desired_addr].is_started
+        return self._clients[desired_addr]
+
+    async def resolve_client(self, addr: Addr) -> Client:
+        desired_host, desired_port = addr[0], str(addr[1])
+
+        # initial connection outside of loop to avoid the sleep penalty
+        conn = await self.new(self.address_resolver.addr)
+        connection_attempts = 1
+
+        advertised_host = conn._server_properties.get("advertised_host")
```

Comment:
- Since we're now using `._server_properties` outside of its class, we should probably make it public (without leading `_`). Alternatively, you can expose specific properties with `@property` decorator
- I think `server_properties["advertised_host"]` would be slightly safer than `server_properties.get("advertised_host")` here, because we're not expecting this properties to be missing. And if they are missing we'd get a more explicit exception